### PR TITLE
👌 IMP: Add LR scheduler and analysis tools

### PR DESCRIPTION
**Pull request by ianagbip1oti**
## Summary by CodeRabbit

* **New Features**
  * Introduced a comprehensive value network analysis tool for inspecting and reporting on network weights, node utilization, feature importance, and bucket differentiation.
  * Expanded data summary reporting with detailed value network feature and king/threat bucket analyses, including new visualizations and statistics.

* **Improvements**
  * Integrated a learning rate scheduler for neural network training, enabling automated and flexible learning rate adjustments.
  * Enhanced policy and value network training with centralized learning rate management and improved progress reporting.
  * Added scalar division support for neural network layers and networks to facilitate gradient scaling.

* **Documentation**
  * Updated contributor guidelines with a new section on communication style.
  * Added detailed experimental reports on king bucketing strategies and their impact on network performance.

* **Bug Fixes**
  * Improved king bucket mapping logic for more consistent feature extraction.

* **Refactor**
  * Refactored optimizer and training code to support generic learning rate schedulers and removed manual learning rate adjustments.
  * Simplified and modularized code for better maintainability and clarity.

* **Style**
  * Minor code cleanup and comment removals for improved readability.

### CODE REVIEW

File: crates/princhess-train/src/bin/value-net-analysis.rs [hunk 1]
```diff
@@ -0,0 +1,1100 @@
...
+    match piece_idx {
+        0 => "pawn",
+        1 => "knight",
+        2 => "bishop",
+        3 => "rook",
+        4 => "queen",
+        5 => "king",
+        _ => "unknown",
+    }
+}
```
[the remark this answers is not in the document]
Comment (ianagbip1oti) [reply crates/princhess-train/src/bin/value-net-analysis.rs]: Fixed in [68b7d09](https://github.com/princesslana/princhess/pull/441/commits/68b7d0908e31e3f6114beb13ea4dfd85a0142ba2)

File: crates/princhess-train/src/bin/value-net-analysis.rs [hunk 2]
```diff
@@ -0,0 +1,1030 @@
...
+                (*piece_idx, *square_idx, *is_opponent),
+                ratio,
+                bucket_values.clone(),
+            ));
+        }
+    }
+
+    // Sort by differentiation ratio
+    differentiation_ratios
+        .sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap_or(std::cmp::Ordering::Equal));
```
[the remark this answers is not in the document]
Comment (ianagbip1oti) [reply crates/princhess-train/src/bin/value-net-analysis.rs]: Fixed in [352e0a0](https://github.com/princesslana/princhess/pull/441/commits/352e0a051f9849c027cbe9cc21b6aedd0b06131b)

File: crates/princhess-train/src/bin/value-net-analysis.rs [hunk 3]
```diff
@@ -0,0 +1,1041 @@
...
+        .map(|(_, ratio, _)| *ratio)
+        .filter(|r| r.is_finite())
+        .collect();
+
+    if !finite_ratios.is_empty() {
+        let avg_ratio = finite_ratios.iter().sum::<f64>() / finite_ratios.len() as f64;
+        let median_ratio = {
+            let mut sorted = finite_ratios.clone();
+            sorted.sort_by(|a, b| a.partial_cmp(b).unwrap());
+            sorted[sorted.len() / 2]
```
[the remark this answers is not in the document]
Comment (ianagbip1oti) [reply crates/princhess-train/src/bin/value-net-analysis.rs]: Fixed in [352e0a0](https://github.com/princesslana/princhess/pull/441/commits/352e0a051f9849c027cbe9cc21b6aedd0b06131b)
